outlines/processors/tracking.py

```diff
@@ -0,0 +1,521 @@
+
+        Raises
+        ------
+        ValueError
+            If batch size > 1 is provided. The tracking processor currently
+            only supports single-batch processing.
+        """
+        # Enforce single batch processing
+        if logits.shape[0] > 1:
+            raise ValueError(
+                "LogitTrackingProcessor only supports single-batch processing. "
+                f"Got batch size {logits.shape[0]}"
+            )
+        if len(input_ids) > 1:
+            raise ValueError(
+                "LogitTrackingProcessor only supports single-batch processing. "
+                f"Got {len(input_ids)} sequences"
+            )
+
+        # Always store the raw logits as unstructured
+        self.unstructured_logits.append(logits[0].detach().cpu().numpy().copy())
+
+        # Store the actual chosen token ID if available
+        if self._started and len(input_ids[0]) > 1:
+            # Get the last token from the current sequence
+            self.chosen_tokens.append(input_ids[0][-1])
+
+        # If we haven't started tracking yet, do so now.
+        # this will only happen on the first call to process_logits.
+        else:
+            self._started = True
+
+        # Apply structural constraints if we have a processor
+        if self.processor is not None:
+            processed = self.processor.process_logits(input_ids, logits)
+            self.structured_logits.append(processed[0].detach().cpu().numpy().copy())
+            return processed
+
+        # For unconstrained generation, structured = unstructured
+        self.structured_logits.append(logits[0].detach().cpu().numpy().copy())
+        return logits
+
+    def get_probabilities(self, as_matrix: bool = False) -> Dict[str, Union[List[NDArray], NDArray]]:
+        """Get probability distributions computed from stored logits.
+
+        Parameters
+        ----------
+        as_matrix : bool
+            If True, convert probability lists to matrices.
+            Each matrix will have shape (vocab_size, n_positions), i.e.
+            return_value['unstructured'] is a vocab_size x n_positions matrix.
+
+            If False, return a list of n_positions arrays, each array having
+            shape (vocab_size,).
+
+        Returns
+        -------
+        Dict[str, Union[List[NDArray], NDArray]]
+            Contains a dictionary with two keys:
+            - unstructured: Raw probability distributions
+            - structured: Probability distributions after constraints
+            Each can be either a list of arrays or a single matrix
+        """
+        # Convert logits to probabilities
+        unstructured_probs = [
+            torch.softmax(torch.tensor(logits), dim=-1).numpy()
+            for logits in self.unstructured_logits
+        ]
+        structured_probs = [
+            torch.softmax(torch.tensor(logits), dim=-1).numpy()
+            for logits in self.structured_logits
+        ]
+
+        if as_matrix:
```

Comment:
Do we need the `as_matrix` parameter? It seems like a straightforward transformation that users can do themselves should they need it

Reply:
I think we need to make a choice here, what's the downside of returning a NumPy array every time?

Reply:
If we choose one, I think probably using the `as_matrix=True` is the best. I'll update it, thanks!